### ISSUE TITLE
Make HTML files editable inside Defold's script editor

### ISSUE THREAD

**Issue by dima13230** [defold/defold#8126]
**Is your feature request related to a problem? Please describe (REQUIRED):**
Currently, I'm working on the SDK that requires to edit both Lua and HTML files often enough and I always have to look into Lua code to then add corresponding code into HTML file, and for Lua I prefer using Defold's script editor. However, I cannot edit HTML files inside Defold so I have to open and keep opened window of another text editor specifically for HTML, which I find quite inconvenient.

**Describe the solution you'd like (REQUIRED):**
I would like to be able to open and edit HTML files inside Defold's editor

**Describe alternatives you've considered (REQUIRED):**
The only alternative I can think of is to nevermind it and continue using external editor for HTML :smile: 

**Additional context (OPTIONAL):**
I have already started working on PR for this feature:
![screenshot](https://github.com/defold/defold/assets/12088459/f574d7a2-2479-40d8-825d-7f5485e42f60)

The main part is done, I'm currently working on syntax highlighting. Mentioning that, I have a little problem implementing it.
Here's my HTML grammar code:

```clojure
(ns editor.code.lang.html
  (:require [editor.code.data :as data])
  (:import [java.io PushbackReader]))

(set! *warn-on-reflection* true)
(set! *unchecked-math* :warn-on-boxed)

(def grammar
  {:name "HTML"
   :scope-name "source.html"
   ;; indent patterns shamelessly stolen from textmate:
   ;; https://github.com/textmate/html.tmbundle/blob/master/Preferences/Miscellaneous.plist
   :indent {
            :end #"(?x)
		^\s*
		(&lt;/(?!html)
		  [A-Za-z0-9]+\b[^&gt;]*&gt;
		|--&gt;
		|&lt;\?(php)?\s+(else(if)?|end(if|for(each)?|while))
		|\}
		)"
            }
   :patterns [
              {:match #"<[^>]*>"
               :name "tag.language.html"}
              {:match #"\b(?:true|false|null)\b"
               :name "constant.language.html"}
              {:match #"(?x)-?(?:0|[1-9]\d*)(?:\n(?:\n\.\d+)?(?:[eE][+-]?\d+)?)?"
               :name "constant.numeric.html"}
              {:begin #"\""
               :begin-captures {0 {:name "punctuation.definition.string.begin.html"}}
               :end #"\""
               :end-captures {0 {:name "punctuation.definition.string.end.html"}}
               :name "string.quoted.double.html"}
              {:begin #"<"
               :begin-captures {0 {:name "punctuation.definition.tag.begin.html"}}
               :end #">"
               :end-captures {0 {:name "punctuation.definition.tag.end.html"}}
               :name "tag.html"}
               ]})

(defn lines->html [lines & options]
  (with-open [lines-reader (data/lines-reader lines)
              pushback-reader (PushbackReader. lines-reader)]))

```
There are two parts in which I attempted to make tag highlighting:
```clojure
{:match #"<[^>]*>"
               :name "tag.language.html"}
```
and
```clojure
{:begin #"<"
               :begin-captures {0 {:name "punctuation.definition.tag.begin.html"}}
               :end #">"
               :end-captures {0 {:name "punctuation.definition.tag.end.html"}}
               :name "tag.html"}
```
One that uses match by Regex string doesn't seem to work at all (nothing is highlighted) and the other one only highlights the tag < and > symbols. I assume I have to add color rules for HTML. If that's the case, where can I do that or what could be the problem?
P.S. I tested aforementioned problematic grammar rules separately since they don't work together and aren't supposed to.

**Comment by dima13230** [comment]
> The color substrings we match are in `editor/code/view.clj`, in the `code-color-scheme` `def`. However, we already colorize `string` and `punctuation`, so I would expect you to see some colors at least. 🤔
> 
> We typically base our grammars on VSCode or TextMate grammars, manually porting them to Clojure maps. We do not support all features of the TextMate grammars yet, though. For example, we don't support repositories.
> 
> I guess it is possible we are missing some feature required to match named start and end tags, but I'd still expect the `strings` to be colorized.

Yes, I'm seeing numbers and strings colored everywhere except inside tags (it means that Regex works correctly and de-colorizes whole content inside tags)
![image](https://github.com/defold/defold/assets/12088459/705bb080-bfcd-46ec-9bc6-ef6e2685ed7f)

Thanks for the info where to look for color schemes, I'll try that!

**Comment by dima13230** [comment]
**Confirmed**: absence of values in `code-color-scheme` was the reason why it wasn't highlighted, after I added color for `tag`, tags are now highlighted and anything inside them (including strings). I guess I'd have to write patterns for inline strings such as values of attributes to fix this, will try to look how they did that in VSCode.
![image](https://github.com/defold/defold/assets/12088459/5efa2ef0-a8bb-4f51-8ec4-bad875bbe23a)

**Comment by ViTalityGH** [comment]
this is necessary for me, too.

**Comment by dima13230** [comment]
I've been working on syntax highlighting, however I didn't check and notice more significant problem: I cannot save edited HTML files, just nothing happens when I press Ctrl + S or attempt to save it by menus. The tab still has *, so it's sure marked as edited. What could be the cause of this?

**Comment by matgis** [comment]
Without knowing what you changed, it's hard to tell.

The `load-registered-resource-node` function in `defold_project.clj` connects the `:save-data` output of the `ResourceNode` to the `Project` node defined in the same file. This is what determines which files are saved when you press Ctrl + S.

You can see that it only makes this connection if the resource type has a `:write-fn`, so maybe that's the problem?

For the change you're attempting, I would make the `HTML` node inherit from `code.resource/CodeEditorResourceNode` and use the `code.resource/register-code-resource-type` function in `html/register-resource-types`. It specifies the `:write-fn` for you.

**Comment by dima13230** [comment]
> Without knowing what you changed, it's hard to tell.
> 
> The `load-registered-resource-node` function in `defold_project.clj` connects the `:save-data` output of the `ResourceNode` to the `Project` node defined in the same file. This is what determines which files are saved when you press Ctrl + S.
> 
> You can see that it only makes this connection if the resource type has a `:write-fn`, so maybe that's the problem?
> 
> For the change you're attempting, I would make the `HTML` node inherit from `code.resource/CodeEditorResourceNode` and use the `code.resource/register-code-resource-type` function in `html/register-resource-types`. It specifies the `:write-fn` for you.

Thanks, I already made `HTML` node inherit from `code.resource/CodeEditorResourceNode` but didn't use `code.resource/register-code-resource-type` what was the root of the problem. Now, after I used it instead of `editor.workspace/register-resource-type` HTML files can be saved.
